### ISSUE TITLE
support generic protocols in typevar solving

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#1714]

**Comment by ooopus** [comment]
I think this is another instance of #1714 (generic protocol involvement in typevar solving).

When sorting a `list[str]` with `key=len`, ty widens the element type to `Sized`, which then causes false positives when the result is used as `str`.
```python
from __future__ import annotations

import re

def build_regex(forms: list[str]) -> str:
    sorted_forms = sorted(forms, key=len)

    reveal_type(sorted_forms)  # ty infers list[Sized] (unexpected)
    # Any downstream use that requires `str` then fails:
    escaped = [re.escape(f) for f in sorted_forms]  # ty: expected str/bytes, got Sized
    return "|".join(escaped)
```
Expected:
- `sorted_forms` should be inferred as `list[str]` (same element type as the input iterable).

Actual:
- `sorted_forms` is inferred as `list[Sized]`, losing the concrete element type and producing false positives.

**Comment by flying-sheep** [comment]
test case: https://play.ty.dev/1eeb19cd-d753-4158-8dcb-b474c6bc6f10

```py
from typing import Literal
from collections.abc import Mapping
from types import MappingProxyType

m: Mapping[Literal[1], str] = MappingProxyType(
    {1: "y"}
)
```

**Comment by Sillocan** [comment]
Another test case that I think is related: https://play.ty.dev/409a8f60-45b9-497b-a753-f00459137a41

```py
from re import L
from itertools import batched

elements: list[str] = ["a", "b", "c", "d"]
for batch in batched(elements, 1):
    reveal_type(batch)
    for elem in batch:
        reveal_type(elem)
```

**Comment by carljm** [comment]
@Sillocan that last example requires #281, since `itertools.batched` uses a custom `__new__` return type to determine its specialization

**Comment by carljm** [comment]
The case shown by @flying-sheep is fixed by https://github.com/astral-sh/ruff/pull/21902

**Comment by carljm** [comment]
I went through all of the issues that had been closed as duplicate of this one. Many of them are in fact now fixed! I reopened a few where there seems to still be more needed.

With the merge of https://github.com/astral-sh/ruff/pull/21902 I'm inclined to say that we should close this blanket issue and track remaining problems in more narrowly targeted issues. @dcreager please let me know if you disagree and we can reopen!

**Comment by carljm** [comment]
The example shown above by @ooopus is not fixed -- it looks similar to #2572

**Comment by dcreager** [comment]
> The example shown above by [@ooopus](https://github.com/ooopus) is not fixed -- it looks similar to [#2572](https://github.com/astral-sh/ty/issues/2572)

I think that one is (the just-opened) https://github.com/astral-sh/ty/issues/2799
